security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java

```diff
@@ -0,0 +1,226 @@
+	                ']', '|', '\\', '/', ':', ';', '\"', '\'', '<', '>', ',', '.', '?'));
+
+	@PreAuthorize("hasAuthority('Authorities.Client_Password_Policy_UPDATE')")
+	@Override
+	protected Mono<ClientPasswordPolicy> updatableEntity(ClientPasswordPolicy entity) {
+		return super.update(entity);
+	}
+
+	@Override
+	protected Mono<Map<String, Object>> updatableFields(ULong key, Map<String, Object> fields) {
+		return Mono.just(fields);
+	}
+
+	public Mono<Boolean> checkPasswordPolicyExists(ULong clientId) {
+		return this.dao.checkClientPasswordPolicyExists(clientId);
+	}
+
+	public Mono<ClientPasswordPolicy> getPolicyByClientId(ULong clientId) {
+		return this.dao.getByClientId(clientId);
+	}
+
+	public Mono<Boolean> checkAllConditions(ULong clientId, String password) {
+
+		return flatMapMono(
+
+		        () -> this.dao.getByClientId(clientId),
+
+		        passwordPolicy -> passwordPolicy != null ? checkAlphanumericExists(passwordPolicy, password)
+		                : Mono.empty(),
+
+		        (passwordPolicy, isAlphaNumberic) -> passwordPolicy.isAtleastOneSpecialChar()
+		                ? checkInSpecialCharacters(password).map(val -> val && isAlphaNumberic)
+		                : Mono.just(isAlphaNumberic),
+
+		        // re check here
+		        (passwordPolicy, isAlphaNumberic, isSpecial) ->
+
+				!passwordPolicy.isSpacesAllowed() && password.contains(" ") ?
+
+				        Mono.just(false) : Mono.just(isAlphaNumberic && isSpecial),
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace) ->
+				{
+			        if (!isSpace.booleanValue())
+				        return securityMessageResourceService.throwMessage(HttpStatus.PRECONDITION_FAILED,
+				                SecurityMessageResourceService.SPACES_MISSING);
+
+			        String regex = passwordPolicy.getRegex();
+			        return regex != null && !regex.contains("")
+			                ? checkRegexPattern(password, regex)
+			                        .map(val -> val && isAlphaNumberic && isSpecial && isSpace)
+			                : Mono.just(isAlphaNumberic && isSpecial && isSpace);
+
+		        }, (passwordPolicy, isAlphaNumberic, isSpecial, isSpace, isRegex) -> {
+
+			        if (!isRegex.booleanValue())
+				        return securityMessageResourceService.throwMessage(HttpStatus.PRECONDITION_FAILED,
+				                SecurityMessageResourceService.REGEX_MISMATCH);
+
+			        return checkStrengthOfPassword(passwordPolicy, password);
+		        },
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace, isRegex, isLength) ->
+
+				isLength.booleanValue() ? Mono.just(isAlphaNumberic && isSpecial && isSpace && isRegex) : Mono.empty(),
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace, isRegex, isLength, isValid) -> Mono
+		                .just(isAlphaNumberic && isSpecial && isSpace && isRegex && isValid)
+
+		// Add past passwords and history check later in that past passwords service
```

Comment:
Where are we doing the past passwords check?

Reply:
handled in user service when they only request for updating password.